packages/caliper-core/test/worker/tx-observers/prometheus-push-tx-observer.js

```diff
@@ -0,0 +1,226 @@
+            roundIndex: 2,
+            roundLabel: 'myTestRound',
+            workerIndex: 0
+        });
+    });
+
+    it('should update transaction statistics during use', async () => {
+        const options = {
+            pushUrl: 'http://my.url.com'
+        };
+        const prometheusPushTxObserver = PrometheusPushTxObserver.createTxObserver(options, undefined, 0);
+        prometheusPushTxObserver._sendUpdate = sinon.stub();
+        await prometheusPushTxObserver.activate(2, 'myTestRound');
+        prometheusPushTxObserver.txSubmitted(100);
+        prometheusPushTxObserver.txFinished({
+            GetStatus: sinon.stub().returns('success'),
+            GetTimeFinal: sinon.stub().returns(101),
+            GetTimeCreate: sinon.stub().returns(10)
+        });
+
+        prometheusPushTxObserver.counterTxSubmitted.hashMap.should.deep.equal({
+            '': { value: 100, labels: {} }
+        });
+        prometheusPushTxObserver.counterTxFinished.hashMap.should.deep.equal({
+            'final_status:success': {
+                labels: {
+                    'final_status': 'success'
+                },
+                value: 1
+            }
+        });
+    });
+
+    it('should reset all counters on deactivate so that statistics do not bleed into other rounds', async () => {
+        const options = {
+            pushUrl: 'http://my.url.com'
+        };
+        const prometheusPushTxObserver = PrometheusPushTxObserver.createTxObserver(options, undefined, 0);
+        prometheusPushTxObserver._sendUpdate = sinon.stub();
+        await prometheusPushTxObserver.activate(2, 'myTestRound');
+        prometheusPushTxObserver.txSubmitted(100);
+        prometheusPushTxObserver.txFinished(
+            [{
+                GetStatus: sinon.stub().returns('success'),
+                GetTimeFinal: sinon.stub().returns(101),
+                GetTimeCreate: sinon.stub().returns(10)
+            },
+            {
+                GetStatus: sinon.stub().returns('success'),
+                GetTimeFinal: sinon.stub().returns(101),
+                GetTimeCreate: sinon.stub().returns(10)
+            }]
+        );
+
+        prometheusPushTxObserver.counterTxSubmitted.hashMap.should.deep.equal({
+            '': {
+                labels: {},
+                value: 100
+            }
+        });
+        prometheusPushTxObserver.counterTxFinished.hashMap.should.deep.equal({
+            'final_status:success': {
+                labels: {
+                    'final_status': 'success'
+                },
+                value: 2
+            }
+        });
+
+        await prometheusPushTxObserver.deactivate();
+
+        // Because we have stubbed out _sendUpdate, the default labels are not applied here
+        prometheusPushTxObserver.counterTxSubmitted.hashMap.should.deep.equal({
+            '': { value: 0, labels: {} }
+        });
+        prometheusPushTxObserver.counterTxFinished.hashMap.should.deep.equal({});
```

Comment:
1. It's a bit strange that the two counters have different empty states. Is there a reason for that?
2. This comment is for the implementation itself rather than for the test. The stats won't overlap, since upon activating a new round, the basic labels will change, resulting in a separate time series. The current behavior will reset the counters to 0 which will remain there, possibly missing the max value in the end if scraped with unfortunate timing. So maybe we shouldn't reset the counters after the rounds? The label change is enough to isolate the rounds completely.

Reply:
I think some experimenting around here is required. I completely agree that we don't want to throw away data, but i'm curious about the implications for this once benchmarks have been completed. With repeat benchmarks using dockerised workers, would we result in a continuous data line for all rounds ever run?

Reply:
looking at the prom-client tests, the difference is to be expected: https://github.com/siimon/prom-client/blob/master/test/counterTest.js

Reply:
> With repeat benchmarks using dockerised workers, would we result in a continuous data line for all rounds ever run?

Yes, but this is true for any service, not just Caliper. If you want multiple independent runs in the same prometheus instance, then you must add a label that differentiates between them. Otherwise, it continues where it left off (to allow for service restarts). In the typical use cases, you will probably tear down prometheus along with Caliper and Fabric.